core/src/service/file_metadata_service.rs

```diff
@@ -1,140 +1,392 @@
+                            path: file.path.to_string(),
+                            updated_at: version,
+                            version: version,
+                            status: Status::Local,
+                        },
+                    )
+                    .unwrap();
+                    debug("CREATE -- SUCCESS".to_string())
                 }
+                Err(err) => error(format!("CREATE -- FAILURE: {:?}", err)),
+            }
+        });
+        let mut updates_local = FileMetadataDb::get_all(db)?;
+        updates_local.retain(|meta| meta.status == Status::Local);
+        let updates_local_res = updates_local
+            .iter()
+            .map(|file| -> Result<FileMetadata, Error> {
+                let content = serde_json::to_string(&FileDb::get(db, &file.id)?)?;
+                let new_version = ApiClient::change_file(&ChangeFileContentRequest {
+                    username: account.username.to_string(),
+                    auth: "JUNK_AUTH".to_string(),
+                    file_id: file.id.to_string(),
+                    old_file_version: file.version,
+                    new_file_content: content,
+                })?;
+                Ok(FileMetadataDb::update(
+                    db,
+                    &FileMetadata {
+                        id: file.id.to_string(),
+                        name: file.name.to_string(),
+                        path: file.path.to_string(),
+                        updated_at: file.updated_at,
+                        version: new_version,
+                        status: Status::Synced,
+                    },
+                )?)
+            });
+        let errors_local = updates_local_res
+            .into_iter()
+            .filter_map(Result::err)
+            .collect::<Vec<Error>>();
+        error(format!("Local Errors: {:?}", errors_local));
+        let updates_remote_res = updates_remote
+            .iter()
+            .map(|meta| -> Result<FileMetadata, Error> {
+                let content = ApiClient::get_file(&GetFileRequest {
+                    file_id: meta.file_id.to_string(),
+                })?;
+                FileDb::update(db, &meta.file_id, &content)?;
+                Ok(FileMetadataDb::update(
+                    db,
+                    &FileMetadata {
+                        id: meta.file_id.to_string(),
+                        name: meta.file_name.to_string(),
+                        path: meta.file_path.to_string(),
+                        updated_at: meta.file_metadata_version,
+                        version: meta.file_content_version,
+                        status: Status::Synced,
+                    },
+                )?)
             });
+        let errors_remote = updates_remote_res
+            .into_iter()
+            .filter_map(Result::err)
+            .collect::<Vec<Error>>();
+        error(format!("Remote Errors: {:?}", errors_remote));
+        Ok(FileMetadataDb::get_all(&db)?)
```

Comment:
I think this overall strategy is good, and conducive to being called periodically. I think we could do a bit of localized cleanup.

Areas of improvement:
Once the client's are back to returning their specific error messages, `update`'s error scenarios is the type of thing we'd want to handle.

They're pretty well "documented/coded" by @tvanderstad, we should probably match on each api call and handle the failures we can. For example:

`SendFailed` let's retry? 
`EditConflict` there are diffs to resolve. 

But I think the overall structure is good, this is all the happy paths, and I'm okay with adding the failure condition in subsequent PR's. We might be well served by doing some work in `docs` and thinking out all the paths.

Reply:
Yeah this is a large PR ☝️ is exactly what I said was to be worked on in subsequent PRs